demonstrations/tutorial_pulse_programming101.py

```diff
@@ -0,0 +1,362 @@
+
+omega = 4.8 * jnp.ones(n_wires)
+g = 0.01 * jnp.ones(n_wires-1)
+
+H_D = qml.op_sum(*[qml.s_prod(omega[i], ad(i) @ a(i)) for i in range(n_wires)])
+H_D += qml.op_sum(*[qml.s_prod(g[i], ad(i) @ a(i+1)) for i in range(n_wires-1)])
+
+##############################################################################
+# The system is driven under the control term
+#
+# .. math:: H_C(t) = \sum_q \Omega_q(t) \left(e^{i\nu_q t} a_q + e^{-i\nu_q t} a^\dagger_q \right)
+# 
+# with the (real) time-dependent amplitude :math:`\Omega(t)` and frequency :math:`\nu_q` of the drive.
+# We let :math:`\Omega(t)` be a piece-wise-constant real function that is optimized alongside the frequencies :math:`\nu_q`.
+
+# TODO use official convenience functions once merged
+def pwc(t1, t2):
+    def wrapped(params, t):
+        N = len(params)
+        idx = jnp.array(N/(t2 - t1) * (t - t1), dtype=int) # corresponding sample
+        return params[idx]
+
+    return wrapped
+
+def envelope(t1, t2, sign=1.):
+    # assuming p = (len(t_bins) + 1) for the frequency nu
+    def wrapped(p, t):
+        return pwc(t1, t2)(p[:-1], t) * jnp.exp(sign*1j*p[-1]*t)
+    return wrapped
+
+duration = 15.
+
+fs = [envelope(0., duration, 1.) for i in range(n_wires)]
+fs += [envelope(0., duration, -1.) for i in range(n_wires)]
+ops = [a(i) for i in range(n_wires)]
+ops += [ad(i) for i in range(n_wires)]
+
+H_C = qml.ops.dot(fs, ops)
+
+##############################################################################
+# Overall, we end up with the time-dependent parametrized Hamiltonian :math:`H(p, t) = H_D + H_C(p, t)`
+# under which the system is evolved for the given time window.
+
+H_pulse = H_D + H_C
+
+dev = qml.device("default.qubit", wires=range(n_wires))
+
+@qml.qnode(dev, interface="jax")
+def qnode(p, t=duration):
+    qml.BasisState(np.array([1, 1, 1, 0, 0, 0]), wires=H_obj.wires)
+    qml.evolve(H_pulse)(params=(*p, *p), t=t)
+    return qml.expval(H_obj)
+
+##############################################################################
+# Our aim is to minimize the energy expectation value :math:`\langle H_\text{obj} \rangle` but we also need to take 
+# certain physical constraints into account. For example, the amplitude of the physical driving field :math:`\Omega(t)` cannot 
+# reach arbitrarily large values and cannot change arbitrarily quickly in time. We therefore add two penalty terms quantifying the 
+# mean absolute square of the amplitude and changes in the amplitude. The overall cost function is thus
+# 
+# .. math:: \mathcal{C}(p) = \langle H_\text{obj} \rangle + \sum_{ki} |p^k_i|^2 + \sum_{ki} |p^k_i - p^k_{i-1}|^2.
+#
+
+def abs_diff(p):
+    """compute |p_i - p_i-1|^2"""
+    return jnp.mean(jnp.abs(jnp.diff(p[:, :-1], axis=1))**2)
+
+def cost_fn(params):
+    # params.shape = (n_wires, {n_bins}+1)
+    p = params[:, :-1]
+    C_exp = qnode(params)           # expectation value
+    C_par = jnp.mean(jnp.abs(p)**2) # parameter values
```

Comment:
It's just a prefactor for the penalty terms, but the equation above suggests that we sum, not average, the amplitude/difference squares.
```suggestion
    return jnp.sum(jnp.abs(jnp.diff(p[:, :-1], axis=1))**2)

def cost_fn(params):
    # params.shape = (n_wires, {n_bins}+1)
    p = params[:, :-1]
    C_exp = qnode(params)           # expectation value
    C_par = jnp.sum(jnp.abs(p)**2) # parameter values
```

Reply:
I will change the formula instead, it makes a big difference!